flair/models/dependency_parser_model.py

```diff
@@ -0,0 +1,624 @@
+                for token_index, (token, tag, head_id) in enumerate(zip(sentence.tokens, sent_tags, sent_arcs)):
+                    token.add_tag(self.tag_type,
+                                  tag,
+                                  score_rel[sentnce_index][token_index])
+                    
+                    token.head_id = int(head_id)
+
+                if print_tree:
+                    tree_printer(sentence, self.tag_type)
+                    print("-" * 50)
+            store_embeddings(batch, storage_mode=embedding_storage_mode)
+    
+    def evaluate(
+            self,
+            data_points: Union[List[DataPoint], Dataset],
+            gold_label_type: str,
+            out_path: Union[str, Path] = None,
+            embedding_storage_mode: str = "none",
+            mini_batch_size: int = 32,
+            num_workers: int = 8,
+            main_evaluation_metric: Tuple[str, str] = ("micro avg", "f1-score"),
+            gold_label_dictionary: Optional[Dictionary] = None,
+        ) -> Result:
+        
+        if not isinstance(data_points, Dataset):
+            data_points = SentenceDataset(data_points)
+        data_loader = DataLoader(data_points,
+                                 batch_size=mini_batch_size,
+                                 num_workers=num_workers)
+
+        lines: List[str] = []
+
+        average_over = 0
+        eval_loss_arc = 0
+        eval_loss_rel = 0
+
+        y_true = []
+        y_pred = []
+
+        parsing_metric = ParsingMetric()
+
+        for batch in data_loader:
+            average_over += 1
+            with torch.no_grad():
+                score_arc, score_rel = self.forward(batch)
+                loss_arc, loss_rel = self._calculate_loss(score_arc, score_rel, batch)
+                arc_prediction, relation_prediction = self._obtain_labels_(score_arc, score_rel)
+                
+            parsing_metric(arc_prediction, relation_prediction, batch, gold_label_type)
+            
+            eval_loss_arc += loss_arc
+            eval_loss_rel += loss_rel
+
+            for (sentence, arcs, sent_tags) in zip(batch, arc_prediction, relation_prediction):
+                for (token, arc, tag) in zip(sentence.tokens, arcs, sent_tags):
+                    token: Token = token
+                    token.add_tag_label("predicted", Label(tag))
+                    token.add_tag_label("predicted_head_id", Label(str(arc)))
+
+                    # append both to file for evaluation
+                    eval_line = "{} {} {} {} {}\n".format(token.text,
+                                                          token.get_tag(gold_label_type).value,
+                                                          str(token.head_id),
+                                                          tag,
+                                                          str(arc))
```

Comment:
this is sometimes a tensor and sometimes an int, leading to uneven printouts

Reply:
Thanks, It's fixed on the recent commit. Also, I add CSV header to prevent confusion. it's better now.


![Screenshot from 2021-11-28 11-38-57](https://user-images.githubusercontent.com/68088136/143762757-9882772f-b1e6-4e90-8c0a-cdd1737f37a7.png)